zookeeper/settings.sls

```diff
@@ -67,14 +67,38 @@
 {%- set hosts_target         = g.get('hosts_target', p.get('hosts_target', 'roles:zookeeper')) %}
 {%- set targeting_method     = g.get('targeting_method', p.get('targeting_method', 'grain')) %}
 
+{%- set cluster_id     = g.get('cluster_id', p.get('cluster_id', None)) %}
 {%- set zookeepers_with_ids = [] %}
 {%- set zookeepers          = [] %}
-{%- set myid_dist           = [] %}
+{%- set myid_dist            = [] %}
 {%- set connection_string   = [] %}
-{%- set minion_ips          = salt['network.ip_addrs']() %}
+{%- set minion_ids          = [salt['network.get_hostname'](),
+                               grains['id'],
+                               grains['fqdn'],
+                               grains['nodename']] + salt['network.ip_addrs']() %}
 
 {% if p.get('nodes') %}
   {%- set zookeeper_nodes   = p.get('nodes', []) %}
+{% elif p.get('clusters') and cluster_id != None %}
+  {%- set zookeeper_nodes   = p.get('clusters').get(cluster_id, []) %}
+{% elif p.get('clusters') %}
+  {%- set zookeeper_nodes_tmp   = [] %}
+  {%- if p.get('clusters')|length == 1 %}
+    {%- set zookeeper_nodes   = p.get('clusters').values()[0] %}
+  {%- else %}
+    {% for cluster, nodes in p.get('clusters').iteritems() %}
+      {%- for node in nodes %}
+        {%- if node in minion_ids %}
+          {%- do zookeeper_nodes_tmp.append(nodes)  %}
+          {%- break %}
+        {%- endif %}
+      {%- endfor %}
+      {%- if zookeeper_nodes_tmp|length != 0 %} 
+        {%- break %}
+      {%- endif %}
+    {% endfor %}
+    {%- set zookeeper_nodes   = zookeeper_nodes_tmp[0] if zookeeper_nodes_tmp|length !=0 else [] %}  
+  {%- endif %}
```

Comment:
You've already verified that Pillar dict has value under the `clusters` key, so no need to use the `get` method again.
Regarding `dictsort` filter, it's always better to avoid "Pythonisms" in Jinja whenever possible, they are unreliable and can lead to weird-looking stack traces. For example, `iteritems()` is not present in Python 3, so the template will fail if someone runs Salt with py3. Also, it returns list of `(key, value)` tuples sorted in unpredictable order.